File: src/main/java/nl/sourcelabs/demos/cf/CfApplication.java
package nl.sourcelabs.demos.cf;

import io.micrometer.core.instrument.MeterRegistry;
import io.micrometer.core.instrument.binder.jvm.ExecutorServiceMetrics;
import org.apache.http.client.HttpClient;
import org.apache.http.client.config.RequestConfig;
import org.apache.http.impl.client.HttpClientBuilder;
import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.context.annotation.Bean;

import java.util.concurrent.Executor;
import java.util.concurrent.Executors;

@SpringBootApplication
public class CfApplication {

  public static void main(String[] args) {
    SpringApplication.run(CfApplication.class, args);
  }

  @Bean
  public Executor theExecutor(final MeterRegistry registry) {
    return ExecutorServiceMetrics
            .monitor(registry, Executors.newFixedThreadPool(10), "myExecutorPool");
  }

  @Bean
  public HttpClient myHttpClient() {
    HttpClientBuilder httpClientBuilder = HttpClientBuilder.create()
            .setDefaultRequestConfig(
                    RequestConfig.custom()
                            .setConnectTimeout(1000)
                            .setSocketTimeout(1000)
                            .build())
            .setMaxConnTotal(100)
            .setMaxConnPerRoute(50);
    return httpClientBuilder.build();
  }

}

File: src/main/java/nl/sourcelabs/demos/cf/CfApplication.kt
package nl.sourcelabs.demos.cf

import org.apache.http.client.config.RequestConfig
import org.apache.http.impl.client.CloseableHttpClient
import org.apache.http.impl.client.HttpClientBuilder
import org.springframework.boot.SpringApplication
import org.springframework.boot.autoconfigure.SpringBootApplication
import org.springframework.boot.web.client.RestTemplateBuilder
import org.springframework.context.annotation.Bean
import org.springframework.http.client.HttpComponentsClientHttpRequestFactory
import org.springframework.web.client.RestTemplate

@SpringBootApplication
class CfApplication {

    @Bean
    fun myHttpClient(): CloseableHttpClient {
        val httpClientBuilder: HttpClientBuilder = HttpClientBuilder.create()
            .setDefaultRequestConfig(
                RequestConfig.custom()
                    .setConnectTimeout(2000)
                    .setSocketTimeout(2000)
                    .build())
            .setMaxConnTotal(100)
            .setMaxConnPerRoute(100)
        return httpClientBuilder.build()
    }

    @Bean
    fun myTemplate(): RestTemplate {
        return RestTemplateBuilder()
            .requestFactory { HttpComponentsClientHttpRequestFactory(myHttpClient()) }
            .rootUri("http://localhost:1080")
            .build()
    }
}

fun main(args: Array<String>) {
    SpringApplication.run(CfApplication::class.java, *args)
}